File: Labs/week3/C_stack_pop_sequence.py
in_str, out_str = input().strip().split()
sz = len(in_str)
stk = []
ans = ""
i = j = 0
flag = True

while True:
    if i == sz and j == sz:
        break
    if not stk:
        stk.append(in_str[i])
        ans += 'P'
        i += 1
    elif stk[-1] == out_str[j]:
        stk.pop()
        ans += 'O'
        j += 1
    else:
        if i == sz:
            flag = False
            break
        stk.append(in_str[i])
        ans += 'P'
        i += 1

if flag:
    print("right")
    print(ans)
else:
    print("wrong")
    print(''.join(stk))

File: Labs/week3/C_stack_pop_sequence.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    string in, out, ans = "";
    cin >> in >> out;
    int sz = in.size();
    stack<char> stk;
    bool flag = true;
    int i = 0, j = 0;
    while (true) {
        if (i == sz && j == sz) break;
        if (stk.empty()) {
            stk.push(in[i]);
            ans += 'P';
            i++;
        } else if (stk.top() == out[j]) {
            stk.pop();
            ans += 'O';
            j++;
        } else {
            if (i == sz) {
                flag = false;
                break;
            }
            stk.push(in[i]);
            ans += 'P';
            i++;
        }
    }
    if (flag) {
        cout << "right" << endl;
        cout << ans << endl;
    } else {
        cout << "wrong" << endl;
        string res = "";
        while (!stk.empty()) {
            res = stk.top() + res;
            stk.pop();
        }
        cout << res << endl;
    }
    return 0;
}
